planner/core/rule_generate_column_substitute.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"context"
+
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/types"
+)
+
+type gcSubstituter struct {
+}
+
+// ExprColumnMap is used to store all expressions of indexed generated columns in a table,
+// and map them to the generated columns,
+// thus we can substitute the expression in a query to an indexed generated column.
+type ExprColumnMap map[expression.Expression]*expression.Column
+
+// optimize try to replace the expression to indexed virtual generate column in where, group by, order by, and field clause
+// so that we can use the index on expression.
+// For example: select a+1 from t order by a+1, with a virtual generate column c as (a+1) and
+// an index on c. We need to replace a+1 with c so that we can use the index on c.
+// See also https://dev.mysql.com/doc/refman/8.0/en/generated-column-index-optimizations.html
+func (gc *gcSubstituter) optimize(ctx context.Context, lp LogicalPlan) (LogicalPlan, error) {
+	exprToColumn := make(ExprColumnMap)
+	collectGenerateColumn(lp, exprToColumn)
+	if len(exprToColumn) == 0 {
+		return lp, nil
+	}
+	return gc.substitute(ctx, lp, exprToColumn), nil
+}
+
+// collectGenerateColumn collect the generate column and save them to a map from their expressions to themselves.
+// For the sake of simplicity, we don't collect the stored generate column because we can't get their expressions directly.
+// TODO: support stored generate column.
+func collectGenerateColumn(lp LogicalPlan, exprToColumn ExprColumnMap) {
+	if ds, ok := lp.(*DataSource); ok {
+		tblInfo := ds.tableInfo
+		for _, idx := range tblInfo.Indices {
+			for _, idxPart := range idx.Columns {
+				colInfo := tblInfo.Columns[idxPart.Offset]
+				if colInfo.IsGenerated() && !colInfo.GeneratedStored {
+					s := ds.schema.Columns
+					col := expression.ColInfo2Col(s, colInfo)
+					if col != nil {
+						exprToColumn[col.VirtualExpr] = col
+					}
+				}
+			}
+		}
+		return
+	}
+	for _, child := range lp.Children() {
+		collectGenerateColumn(child, exprToColumn)
+	}
+}
+
+func (gc *gcSubstituter) substitute(ctx context.Context, lp LogicalPlan, exprToColumn ExprColumnMap) LogicalPlan {
+	sctx := lp.SCtx().GetSessionVars().StmtCtx
+	var expr *expression.Expression
+	var tp types.EvalType
+	switch x := lp.(type) {
+	case *LogicalSelection:
+		for _, cond := range x.Conditions {
+			sf, ok := cond.(*expression.ScalarFunction)
```

Comment:
Can you try to wrap the codes below in a separate function, like `substituteCondition()`?